### ISSUE TITLE
<all>: Microsoft clang doesn't propogate correct alignment for vector extensions to STL

### ISSUE THREAD

**Issue by alecazam** [microsoft/STL#3608]
**Describe the bug**
We have a simd library based on the gcc/clang extensions.  And it doesn't fail on 4 other platforms, but it does fail on Microsoft clang when using Microsoft STL.   We had to add an __aligned__(16) into the vector, or else it was getting stuck on 8B boundaries when used inside a std::array, std::unordered_array, and std::vector.  

Here's a failing text case of the basic source code.  We have exceptions and RTTI off.  I have the following off too that affect the STL.  It's a debug x64 build with typical compile settings.

CXXFLAGS += -D_ITERATOR_DEBUG_LEVEL=0 -D_D_HAS_EXCEPTIONS=0

**Command-line test case**
```
#include <array>

// This should be aligned to 16B, but isn't below.  It's using 8B align.
typedef __attribute__((__ext_vector_type__(3))) float my_simd_float3;

// This works.
//typedef __attribute__((__ext_vector_type__(3),  __aligned__(16))) float my_simd_float3;

struct TestStruct
{
	my_simd_float3 v0 = { 1,0,0 };
};

class TestStruct1
{
public:
	// getting rid of this vtable works, since 8B alignment isn't there.
	virtual ~TestStruct1() {}
	std::array< TestStruct, 10 > testArray;
};

intmain()
{
	// This works.  Also makes case below not fail, but simply due to realignment.
	// TestStruct array[10] = {}; 

	// here's the crash repro, Microsoft STL isn't getting alignment
	TestStruct1 test;

   return 0;
}


```

**Expected behavior**
Expected that this code would be aligned to 16B like it requires.  Otherwise we get simd crashes from data aligned to 8B.   It's like it stuffs the vtable ptr into the class, and then immediately packs the simd vector (or matrix) after that.  That's not supposed to happen.

**STL version**
    Microsoft Visual Studio Professional 2019 (2)
   Version 16.11.16
   C++14 build

Microsoft Visual Studio Professional 2019 (2)
Version 16.11.16
VisualStudio.16.Release/16.11.16+32602.291
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Professional

Visual C++ 2019   00435-30023-55627-AA321
Microsoft Visual C++ 2019

ASP.NET and Web Tools 2019   16.11.106.23128
ASP.NET and Web Tools 2019

Azure App Service Tools v3.0.0   16.11.106.23128
Azure App Service Tools v3.0.0

BuildWithFASTBuildCommand Extension   1.0
BuildWithFASTBuildCommand Visual Studio Extension Detailed Info

C# Tools   3.11.0-4.22108.8+d9bef045c4362fbcab27ef35daec4e95c8ff47e1
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

CodeLensOopProviderCommands Extension   1.0
CodeLensOopProviderCommands Visual Studio Extension Detailed Info

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

GnD Editor Extension   1.8.0.0
GnD Editor Extension
Copyright (C) 2022 Sony Interactive Entertainment Inc. All rights reserved.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

NuGet Package Manager   5.11.2
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

OptionsPagePackage Extension   1.0
OptionsPagePackage Visual Studio Extension Detailed Info

P4VS - Helix Plugin for Visual Studio   2021.2.241.3698
P4VS is the Helix Plugin for Microsoft Visual Studio. P4VS provides developers working within Visual Studio access to Helix version control features, including check in/check out, view file history, and more.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   16.0.30526.2002
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.11.0-4.22108.8+d9bef045c4362fbcab27ef35daec4e95c8ff47e1
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

    ``

**Comment by alecazam** [comment]
We haven't yet moved to VS2022, and I'm sure there's a new clang there to try.  So maybe this is fixed in that version.

**Comment by frederick-vs-ja** [comment]
This seems only related to Clang. Perhaps you should report the bug at [LLVM repo](https://github.com/llvm/llvm-project/issues).

Have you tested replacing `std::array` with the following theoretically isomorphic thing?
```C++
template<class T, std::size_t N>
struct MyArray {
    T elem[N];
};
```

If this bug can also be triggered by `MyArray<TestStruct, 10>`, then it should be unrelated to MSVC STL.

**Comment by alecazam** [comment]
The code above has a raw array that works fine.  It’s when switching to std::array or unordered_map of a struct with that ext_vector_type that things fail.  So it’s the STL.  We use similar clang with different platform STL on 5 other platforms that don’t crash or get the alignment wrong.  Clang gens the correct code for struct/class usage not in STL containers.

**Comment by alecazam** [comment]
Also note this isn't a stack vs. heap alignment issue.  It occurs in the example code on the stack, and in our code with a map that is allocating elements on the heap.

**Comment by StephanTLavavej** [comment]
I agree with @frederick-vs-ja. Note that the request isn't to try a *raw* array `T[N]`. It's to compare `MyArray<T, N>`, which is an aggregate. `std::array` is the same kind of aggregate (no constructors, just member functions), so the compiler should handle `MyArray` and `std::array` identically. I believe that this is purely a Clang compiler bug with no STL involvement, as `std::array` is such a simple type in this respect.

> And this compiles/runs fine.
> `MyArray<TestStruct, 10> test = {};`

Can you please try to replace `std::array` in `TestStruct1` in the original repro, so the entire setup is identical except for this one change? I observe that it's default-initializing `TestStruct1 test;` which then has a data member, whereas `MyArray<TestStruct, 10> test = {};` is list-initializing the thing directly (instead of being a data member).

**Comment by alecazam** [comment]
That works fine too.  Like I mentioned, we're only seeing this with structs with simd content held in an STL container/template.   And we don't have this issue with clang and different STL implementation on all our other platforms.  It's like the alignment to ext_vector_type isn't honored.

```
class TestStruct1
{
public:
	// getting rid of this vtable works, since 8B alignment isn't there.
	virtual ~TestStruct1() {}
	
        // This crashes 8B offset after vtable
	//std::array< TestStruct, 10 > testArray;
	 
        // This works 16B offset
	MyArray<TestStruct, 10> testArray;

        // This works
        // TestStruct testArray[10];
};
```

**Comment by AlexGuteniev** [comment]
The reason of different behavior is `#pragma pack(push, _CRT_PACKING)` / `#pragma pack(pop)` around the `std::array`.

Elimination of them would be an ABI break, so it is at least `vNext`. 
I'm not sure if even it is wanted for `vNext`, possibly even falls into non-goals.

@StephanTLavavej , thoughts?

**Comment by StephanTLavavej** [comment]
Thanks @AlexGuteniev, that explains it.

I haven't looked into this deeply (I am severely overloaded at the moment) but this seems like a Clang bug - I think it might be treating the pragma packs differently from MSVC. My understanding for MSVC is that the pragma pack dance defends against "clever" users who have passed `/Zp1` requesting a small packing - the pragmas say "no, we're going to perform layout of STL types normally with the default packing, even if you want your own types to be weird". However, the pragmas should *never* be capable of giving highly aligned types *less* packing than they want.

Maybe the issue is that this `__attribute__((__ext_vector_type__(3)))` stuff doesn't mark the types as "highly aligned" for Clang, the way that a `__declspec(align)` / `alignas` would for MSVC? (Again, I have not looked into this deeply, but I think these words make sense in this order.)
